File: pcm_client.rb
#!/usr/bin/env ruby
# pcm_client.rb

require 'socket'

class PCMClient
  HOST  = 'localhost'
  PORT  = 5016
  CHUNK = 4096          # read buffer

  def initialize
    @sock = TCPSocket.new(HOST, PORT)
    # Use PulseAudio (WSLg-friendly) instead of ALSA aplay
    @player = IO.popen('paplay --format=s16le --channels=2 --rate=48000 --raw', 'w')
  end

  def stream
    puts '[*] streaming 48 kHz / 16-bit / stereo …  Ctrl-C to stop'
    loop { @player.write(@sock.read(CHUNK)) }
  rescue Interrupt
    puts "\n[*] shutting down"
  ensure
    close
  end

  private

  def close
    @sock.close
    @player.close
  end
end

# --------------------------
PCMClient.new.stream if __FILE__ == $0

File: pcm_client.py
#!/usr/bin/env python3
import socket
import subprocess
import sys
import signal


class PCMClient:
    HOST = 'localhost'
    PORT = 5016
    CHUNK = 4096

    def __init__(self):
        self.sock = None
        self.player = None

    def connect(self):
        self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self.sock.connect((self.HOST, self.PORT))
        
        self.player = subprocess.Popen(
            ['paplay', '--raw', '--format=s16le', '--channels=2', '--rate=48000'],
            stdin=subprocess.PIPE
        )

    def stream(self):
        print('[*] streaming 48 kHz / 16-bit / stereo … Ctrl-C to stop')
        try:
            while True:
                data = self.sock.recv(self.CHUNK)
                if not data:
                    break
                self.player.stdin.write(data)
                self.player.stdin.flush()
        except KeyboardInterrupt:
            print('\n[*] shutting down')
        finally:
            self.close()

    def close(self):
        if self.sock:
            self.sock.close()
        if self.player:
            self.player.stdin.close()
            self.player.terminate()
            self.player.wait()


if __name__ == '__main__':
    client = PCMClient()
    client.connect()
    client.stream()
